lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go

```diff
@@ -89,13 +94,73 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
 		subProto.NetworkId = &protos.NetworkID{Id: networkID}
 		subProtos = append(subProtos, subProto)
 	}
+
+	// Short-circuit subscriber data transmission by comparing cloud & gateway digests.
+	// If digest generation fails, the error is swallowed to not affect the main functionality.
+	//
+	// Note: currently the logic only applies to a request for the first page to avoid concurrency issues;
+	// more to be implemented with gateway subscriber cache that will canonize cloud digests.
+	noUpdates, digest := false, &req.PreviousDigest
+	if req.PageToken == "" {
+		digest, err = getDigest(gateway, apnsByName, apnResourcesByAPN)
+		if err != nil {
+			glog.Errorf("Generating digest for subscribers in network of gateway %s failed", gatewayID)
+		} else {
+			noUpdates = req.PreviousDigest == *digest
+			if noUpdates {
+				subProtos = []*lte_protos.SubscriberData{}
+			}
+		}
+	}
+
 	listRes := &lte_protos.ListSubscribersResponse{
 		Subscribers:   subProtos,
 		NextPageToken: nextToken,
+		Digest:        *digest,
+		NoUpdates:     noUpdates,
 	}
 	return listRes, nil
 }
 
+// getDigest loads all subscribers registered on the current network, and returns
+// a deterministic subscriber flat digest.
+func getDigest(gateway *protos.Identity_Gateway, apnsByName map[string]*lte_models.ApnConfiguration, apnResourcesByAPN lte_models.ApnResources) (*string, error) {
```

Comment:
This line seems to be too long even for Go's generous 120 chars convention.
Also, strings in Go are immutable & are 'pointers' already, like slices. they are passed/returned by values.